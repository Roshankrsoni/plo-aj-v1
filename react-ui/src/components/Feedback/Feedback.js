import React from 'react'
import './Feedback.css'
import axios from 'axios'


class Feedback extends React.Component {
    state = {
        message: '',
        messageSent: false,
        messageFailure: false,
        messageSending: false
    }
    handleChange = e => {
        this.setState({ [e.target.name]: e.target.value })
    }

    sendEmail = (event) => {

        event.preventDefault();
        var { message } = this.state;

        this.setState(() => {
            return {
                messageSending: true
            }
        })
        if (message !== '') {
            axios({
                method: "post",
                url: "https://portfolio-mailserver.herokuapp.com/mail",
                data: {
                    //userName: email,
                    userEmail: 'anoopjadhav@gmail.com',
                    userMessage: "Message : " + message
                },
                headers: {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json',
                }

            }).then((response) => {

                this.setState(() => {
                    return {
                        messageSent: true,
                        messageSending: false
                    }
                })

                if (response.data.msg === 'success') {
                    setTimeout(() => {
                        this.setState(() => {
                            return {
                                messageSent: false
                            }
                        })
                        this.resetForm()
                        this.setState(() => {
                            return {
                                message: ''
                            }
                        })
                    }, 3000)

                } else if (response.data.msg === 'fail') {

                    this.setState(() => {
                        return {
                            messageFailure: true,
                            messageSending: false
                        }
                    })

                    setTimeout(() => {
                        this.setState(() => {
                            return {
                                messageFailure: false
                            }
                        })
                    }, 3000)

                }
            }).catch(err => {
                this.setState(() => {
                    return {
                        messageFailure: true,
                        messageSending: false
                    }
                })

                setTimeout(() => {
                    this.setState(() => {
                        return {
                            messageFailure: false
                        }
                    })
                }, 3000)

            })
        }
    }

    resetForm() {
        document.getElementById('contact-form').reset();
    }
    render() {
        return <div className={"show-on-scroll col-md-7 page-1 text-left section feedback " + this.props.class}>
            <div className="section-title grey4 h2 bold">Feedback</div>
            <form id="contact-form" className="subsection" onSubmit={this.sendEmail.bind(this)} method="POST">
                {this.state.messageSent && <div className="default-text success-banner mb-2">Thank you for your feedback</div>}
                {this.state.messageFailure && <div className="default-text error-banner mb-2">Sorry Couldn't send your message. Please try again later.</div>}
                {this.state.messageSending && <div className="default-text neutral-banner mb-2">Submitting your feedback...</div>}

                <div className="subsection-data">
                    <div className="subsection-title body-text letterspacing-1">Please provide a constructive feedback.</div>
                    <div className="red body-text mt-2">
                        <textarea className="" type="type" name="message" placeholder={this.state.message} onChange={this.handleChange} />
                    </div>

                    <div className="text-left mt-3"><button type="submit">Send</button></div>
                </div>
            </form>
        </div>
    }

}

export default Feedback;
