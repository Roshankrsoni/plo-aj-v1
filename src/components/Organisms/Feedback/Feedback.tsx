import React, { ChangeEventHandler, FormEvent, useState } from "react";
import axiosInstance from "../../../helpers/axios";
import Banner from "../../Atoms/Banner/Banner";

const Feedback = () => {
  const [message, updateMessage] = useState<string>();
  const [bannerStatus, updateBannerStatus] = useState<string>();

  const handleChange: ChangeEventHandler<Element> = (event) => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    updateMessage(event.target.value);
  };

  const sendEmail = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    try {
      if (message) {
        updateBannerStatus("neutral");
        const response = await axiosInstance.post("/mail", {
          userEmail: "anoopjadhav@gmail.com",
          userMessage: "Message : " + message,
        });

        if (response.data.msg === "success") {
          updateBannerStatus("success");
          updateMessage("");
        } else if (response.data.msg === "fail") {
          throw new Error("Response Error");
        }
      } else {
        updateBannerStatus("");
      }
    } catch (e) {
      updateBannerStatus("");
    }
  };

  const showBanner = () => {
    let errMessage;
    switch (bannerStatus) {
      case "success":
        errMessage = "Thank you for your feedback";
        break;
      case "error":
        errMessage =
          "Sorry Couldn't send your message. Please try again later.";
        break;
      case "neutral":
        errMessage = "Submitting your feedback...";
        break;
      default:
        errMessage = "";
        break;
    }

    return (
      errMessage !== "" && (
        <Banner
          data-testid="banner"
          type={bannerStatus}
          text={errMessage}
          closeBanner={() => updateBannerStatus("")}
        />
      )
    );
  };

  return (
    <>
      {showBanner()}
      <div className="col-md-7 page-1 text-left section feedback ">
        <div className="section-title h2 bold">Feedback</div>
        <form
          id="contact-form"
          className="subsection"
          onSubmit={(event) => sendEmail(event)}
          method="POST"
        >
          <div className="subsection-data">
            <label
              htmlFor="feedback-box"
              className="subsection-title body-text letterspacing-1"
            >
              Please provide a constructive feedback.
            </label>
            <div className="red body-text mt-2">
              <textarea
                id="feedback-box"
                data-testid="feedback-box"
                className=""
                name="message"
                value={message}
                placeholder={message}
                onChange={handleChange}
              />
            </div>
            <div className="text-left mt-3">
              <button type="submit" data-testid="feedback-button">
                Send
              </button>
            </div>
          </div>
        </form>
      </div>
    </>
  );
};

export default Feedback;
