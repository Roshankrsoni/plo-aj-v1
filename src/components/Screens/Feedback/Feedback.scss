.feedback-content {
  .subsection-title {
    margin-bottom: 1rem;
    display: block;
  }
  .buttonContainer {
    margin-top: 1rem;
  }
}
.loading .icon svg {
  animation: rotate360 1s linear infinite;
}

@keyframes rotate360 {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
