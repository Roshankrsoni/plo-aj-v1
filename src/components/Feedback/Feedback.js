import React from 'react'
import './Feedback.css'
import axiosInstance from '../axios'


class Feedback extends React.Component {
    state = {
        message: '',
        messageStatus: ''
    }
    handleChange = e => {
        this.setState({ [e.target.name]: e.target.value })
    }

    sendEmail = (event) => {

        event.preventDefault();
        var { message } = this.state;

        this.setState(() => {
            return {
                messageStatus: 'neutral'
            }
        })
        if (message !== '') {
            axiosInstance.post('/mail',{
                userEmail: 'anoopjadhav@gmail.com',
                userMessage: "Message : " + message
            }).then((response) => {

                this.setState(() => {
                    return {
                        messageStatus: 'success'
                    }
                })

                if (response.data.msg === 'success') {
                    setTimeout(() => {
                        this.setState(() => {
                            return {
                                messageStatus: '',
                                message: ''
                            }
                        })
                        this.resetForm()
                    }, 3000)

                } else if (response.data.msg === 'fail') {

                    this.setState(() => {
                        return {
                            messageStatus: 'failed'
                        }
                    })

                    setTimeout(() => {
                        this.setState(() => {
                            return {
                                messageStatus: ''
                            }
                        })
                    }, 3000)

                }
            }).catch(err => {
                this.setState(() => {
                    return {
                        messageStatus: 'failed'
                    }
                })

                setTimeout(() => {
                    this.setState(() => {
                        return {
                            messageStatus: ''
                        }
                    })
                }, 3000)

            })
        }
    }

    resetForm() {
        document.getElementById('contact-form').reset();
    }
    render() {
        return <div className={"show-on-scroll col-md-7 page-1 text-left section feedback " + this.props.class}>
            <div className="section-title grey4 h2 bold">Feedback</div>
            <form id="contact-form" className="subsection" onSubmit={this.sendEmail.bind(this)} method="POST">
                {this.state.messageStatus === 'success' && <div className="default-text success-banner mb-2">Thank you for your feedback</div>}
                {this.state.messageStatus === 'failed'  && <div className="default-text error-banner mb-2">Sorry Couldn't send your message. Please try again later.</div>}
                {this.state.messageStatus === 'neutral'  && <div className="default-text neutral-banner mb-2">Submitting your feedback...</div>}

                <div className="subsection-data">
                    <div className="subsection-title body-text letterspacing-1">Please provide a constructive feedback.</div>
                    <div className="red body-text mt-2">
                        <textarea className="" type="type" name="message" placeholder={this.state.message} onChange={this.handleChange} />
                    </div>

                    <div className="text-left mt-3"><button type="submit">Send</button></div>
                </div>
            </form>
        </div>
    }

}

export default Feedback;
