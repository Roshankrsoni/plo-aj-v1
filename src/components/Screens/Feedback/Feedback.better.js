import React from "react";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

//setup mock server
import { setupServer } from "msw/node";
import { rest } from "msw";

import Store from "../Store/Store";
import { Provider } from "react-redux";
// the component to test
import Feedback from "./Feedback";

const server = setupServer(
  rest.post(
    "https://portfolio-mailserver.herokuapp.com/mail",
    (req, res, ctx) => {
      sessionStorage.setItem("is-authenticated", "true");
      return res(ctx.json({ msg: "success" }));
    }
  )
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("Submit if feedback message is not empty", async () => {
  render(
    <Provider store={Store}>
      <Feedback />
    </Provider>
  );

  const feedbackBox = screen.getByTestId("feedback-box");
  fireEvent.change(feedbackBox, {
    target: { value: "Some Random Feedback Message" },
  });

  fireEvent.click(screen.getByTestId("feedback-button"));

  await waitFor(() =>
    expect(screen.getByTestId("banner")).toHaveTextContent(
      "Submitting your feedback..."
    )
  );

  await waitFor(() =>
    expect(screen.getByTestId("banner")).toHaveTextContent(
      "Thank you for your feedback"
    )
  );
});

test("Do not Submit if feedback message is empty", async () => {
  render(
    <Provider store={Store}>
      <Feedback />
    </Provider>
  );

  const feedbackBox = screen.getByTestId("feedback-box");
  fireEvent.change(feedbackBox, {
    target: { value: "" },
  });

  fireEvent.click(screen.getByTestId("feedback-button"));
  expect(screen.queryByTestId("banner")).not.toBeInTheDocument();
});

test("Do not Submit if Server throws an error", async () => {
  server.use(
    rest.post(
      "https://portfolio-mailserver.herokuapp.com/mail",
      (req, res, ctx) => {
        sessionStorage.setItem("is-authenticated", "true");
        return res.networkError("Failed to connect");
      }
    )
  );
  render(
    <Provider store={Store}>
      <Feedback />
    </Provider>
  );
  const feedbackBox = screen.getByTestId("feedback-box");
  fireEvent.change(feedbackBox, {
    target: { value: "Some value " },
  });

  fireEvent.click(screen.getByTestId("feedback-button"));
  expect(screen.getByTestId("banner")).toHaveTextContent(
    "Submitting your feedback..."
  );
  await waitFor(() => {
    expect(screen.getByTestId("banner")).toHaveTextContent(
      "Sorry Couldn't send your message. Please try again later."
    );
  });

  fireEvent.click(screen.getByTestId("bannerclosebutton"));
  expect(screen.queryByTestId("banner")).not.toBeInTheDocument();
});

test("Do not Submit if Server fails to update feedback", async () => {
  server.use(
    rest.post(
      "https://portfolio-mailserver.herokuapp.com/mail",
      (req, res, ctx) => {
        sessionStorage.setItem("is-authenticated", "true");
        return res(ctx.json({ msg: "fail" }));
      }
    )
  );
  render(
    <Provider store={Store}>
      <Feedback />
    </Provider>
  );
  const feedbackBox = screen.getByTestId("feedback-box");
  fireEvent.change(feedbackBox, {
    target: { value: "Some value " },
  });

  fireEvent.click(screen.getByTestId("feedback-button"));
  expect(screen.getByTestId("banner")).toHaveTextContent(
    "Submitting your feedback..."
  );
  await waitFor(() => {
    expect(screen.getByTestId("banner")).toHaveTextContent(
      "Sorry Couldn't send your message. Please try again later."
    );
  });

  fireEvent.click(screen.getByTestId("bannerclosebutton"));
  expect(screen.queryByTestId("banner")).not.toBeInTheDocument();
});
