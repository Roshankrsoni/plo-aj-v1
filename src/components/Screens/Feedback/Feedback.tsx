import React, { FormEvent, useState } from "react";
import axiosInstance from "../../../helpers/axios";
import Banner, { BannerStatus } from "../../Atoms/Banner/Banner";
import SectionVisibilityHOC from "../../Organisms/SectionVisibilityHOC/SectionVisibilityHOC";
import T from "../../../translations/en_IN";
import Button from "../../Atoms/Button/Button";
import { RiMailSendLine } from "react-icons/ri";
import { useGlobalContext } from "../../../context/GlobalContext";
import { AiOutlineLoading3Quarters } from "react-icons/ai";
import "./Feedback.scss";
import constants from "../../../helpers/constants";
const Feedback = () => {
  const [message, updateMessage] = useState<string>();
  const [bannerStatus, updateBannerStatus] = useState<BannerStatus>("neutral");
  const [showBanner, setShowBanner] = useState(false);

  const { profileData } = useGlobalContext();
  const { contactDetails } = profileData;
  const { emailId } = contactDetails;

  const getBannerMessage = {
    success: T.THANK_YOU_FOR_FEEDBACK,
    error: T.SORRY_COULDNT_SEND_MSG,
    neutral: T.SUBMITTING_FEEDBACK,
  };

  function onFeedbackInputChange(e: React.FormEvent<HTMLTextAreaElement>) {
    updateMessage((e.target as HTMLTextAreaElement).value);
  }

  const sendEmail = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (message) {
      try {
        setShowBanner(true);
        updateBannerStatus("neutral");
        const response = await axiosInstance.post("/mail", {
          userEmail: emailId,
          userMessage: `Message : ${message}`,
        });

        if (response.data.msg === "success") {
          updateBannerStatus("success");
        } else if (response.data.msg === "fail") {
          updateBannerStatus("error");
        }
        updateMessage("");
      } catch (e: unknown) {
        updateMessage("");
        updateBannerStatus("error");
        console.error((e as DOMException).message as unknown as string);
      }
    }
  };

  return (
    <>
      {showBanner && (
        <Banner
          data-testid="banner"
          type={bannerStatus}
          text={getBannerMessage[bannerStatus]}
          closeBanner={() => {
            updateMessage("");
            updateBannerStatus("neutral");
            setShowBanner(false);
          }}
        />
      )}
      <div className="section-title h2 bold">{T.FEEDBACK}</div>
      <form
        id="contact-form"
        className="subsection"
        onSubmit={(event) => sendEmail(event)}
        method="POST"
      >
        <div className="subsection-data">
          <label
            htmlFor="feedback-box"
            className="subsection-title body-text letterspacing-1"
          >
            {T.FEEDBACK_MESSAGE}
          </label>
          <div className="primary-color body-text mt-2">
            <textarea
              id="feedback-box"
              data-testid="feedback-box"
              className=""
              name="message"
              value={message}
              placeholder={message}
              onChange={onFeedbackInputChange}
            />
          </div>
          <div className="text-left mt-3">
            <Button
              className={
                bannerStatus === "neutral" && showBanner ? "loading" : ""
              }
              type="submit"
              testID="feedback-button"
              label={T.SEND}
              Icon={
                bannerStatus === "neutral" && showBanner
                  ? AiOutlineLoading3Quarters
                  : RiMailSendLine
              }
            />
          </div>
        </div>
      </form>
    </>
  );
};

export default SectionVisibilityHOC(Feedback, constants.classNames.FEEDBACK);
