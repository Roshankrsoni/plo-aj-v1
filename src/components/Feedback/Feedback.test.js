import React from 'react';
import { shallow } from 'enzyme';
import { Feedback } from './Feedback';

import dummyData from '../TestData/testData.json'

describe('MyComponent', () => {
    it('should render correctly in "debug" mode', () => {
        const component = shallow(<Feedback dbData={dummyData} />);
        expect(component).toMatchSnapshot();
    });
});