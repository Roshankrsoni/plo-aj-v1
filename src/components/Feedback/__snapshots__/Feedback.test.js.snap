// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent should render correctly in "debug" mode 1`] = `
<Fragment>
  <div
    className="show-on-scroll col-md-7 page-1 text-left section feedback undefined"
  >
    <div
      className="section-title grey4 h2 bold"
    >
      Feedback
    </div>
    <form
      className="subsection"
      id="contact-form"
      method="POST"
      onSubmit={[Function]}
    >
      <div
        className="subsection-data"
      >
        <div
          className="subsection-title body-text letterspacing-1"
        >
          Please provide a constructive feedback.
        </div>
        <div
          className="red body-text mt-2"
        >
          <textarea
            className=""
            name="message"
            onChange={[Function]}
            type="type"
          />
        </div>
        <div
          className="text-left mt-3"
        >
          <button
            type="submit"
          >
            Send
          </button>
        </div>
      </div>
    </form>
  </div>
</Fragment>
`;
